cpp/source/deal.II/new_supercapacitor.templates.h

```diff
@@ -0,0 +1,358 @@
+        if ((cell->face(face)->at_boundary()) &&
+            (cell->face(face)->boundary_id() == cathode_boundary_id))
+          for (unsigned int face_q_point = 0; face_q_point < n_face_q_points;
+               ++face_q_point)
+          {
+            fe_face_values.reinit(cell, face);
+            surface_area += fe_face_values.JxW(face_q_point);
+          }
+
+  // Create the post-processor
+  post_processor_params =
+      std::make_shared<New_SuperCapacitorPostprocessorParameters<dim>>(
+          std::make_shared<boost::property_tree::ptree>(database));
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::inspect(EnergyStorageDeviceInspector *inspector)
+{
+  inspector->inspect(this);
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::print_data(std::ostream &os) const
+{
+  std::ignore = os;
+  throw std::runtime_error("This function is not implemented");
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::get_voltage(double &voltage) const
+{
+  post_processor->get("voltage", voltage);
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::get_current(double &current) const
+{
+  post_processor->get("current", current);
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::reset_voltage(double const voltage)
+{
+  std::ignore = voltage;
+  throw std::runtime_error("This function is not implemented");
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::reset_current(double const current)
+{
+  std::ignore = current;
+  throw std::runtime_error("This function is not implemented");
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::evolve_one_time_step_constant_current(
+    double const time_step, double const current)
+{
+  BOOST_ASSERT_MSG(surface_area != 0., "The surface area is zero.");
+  double const constant_current_density = current / surface_area;
+  bool const rebuild =
+      (electrochemical_physics_params->constant_current_density ==
+       constant_current_density)
+          ? false
+          : true;
+  electrochemical_physics_params->constant_current_density =
+      constant_current_density;
+  evolve_one_time_step(time_step, ConstantCurrent, rebuild);
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::evolve_one_time_step_constant_voltage(
+    double const time_step, double const voltage)
+{
+  bool const rebuild =
+      (electrochemical_physics_params->constant_voltage == voltage) ? false
+                                                                    : true;
+  electrochemical_physics_params->constant_voltage = voltage;
+  evolve_one_time_step(time_step, ConstantVoltage, rebuild);
+}
+
+template <int dim>
+void New_SuperCapacitor<dim>::evolve_one_time_step_constant_power(
+    double const time_step, double const power)
+{
+  BOOST_ASSERT_MSG(surface_area != 0., "The surface area is zero.");
+  dealii::Vector<double> old_solution(solution->block(0));
+  int const max_iterations       = 10;
+  double const percent_tolerance = 1.0e-2;
```

Comment:
probably should make it clear these are related to the Picard iteration and have nothing to do with the solver parameters.